db/collection.go

```diff
@@ -453,3 +585,37 @@ func setNewInstanceID(t []byte) (core.InstanceID, []byte) {
 	}
 	return newID, patchedValue
 }
+
+func compileJSFunc(v []byte, args ...string) ([]byte, error) {
+	if v != nil && len(v) == 0 {
+		v = nil
+	}
+	if v == nil {
+		return nil, nil
+	}
+	script := fmt.Sprintf(`function _fn(%s) {%s}`, strings.Join(args, ","), string(v))
+	script = strings.Replace(script, "\t", "", -1)
+	if _, err := goja.Compile("", script, true); err != nil {
+		return nil, fmt.Errorf("error compiling js func: %v", err)
+	}
+	return []byte(script), nil
+}
+
+func getJSFunc(vm *goja.Runtime, obj []byte, id thread.PubKey) (goja.Callable, goja.Value, error) {
+	_, err := vm.RunString(string(obj))
+	if err != nil {
+		return nil, nil, err
+	}
+	fn, ok := goja.AssertFunction(vm.Get("_fn"))
+	if !ok {
+		return nil, nil, fmt.Errorf("object is not a function: %s", string(obj))
+	}
+	var idv goja.Value
+	if id != nil {
+		idv, err = vm.RunString(fmt.Sprintf("'%s'", id.String()))
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	return fn, idv, nil
```

Comment:
Nice. So we want to call a func, all parameters should be evaluated and passed as "object" rather than values.

Reply:
Yep, function params have to be evaluated as `goja.Value` before they can be understood internally.